src/ja/common/docker_context.py

```diff
@@ -0,0 +1,85 @@
+'''
+Contains definitions of the docker context of a job.
+'''
+from abc import ABC, abstractmethod
+from typing import List
+
+
+class MountPoint:
+    '''
+    A mount point consists of:
+    1. A directory to be mounted.
+    2. The target path in the Docker container where the directory needs to be
+    mounted.
+    '''
+    def __init__(self, source: str, target: str):
+        '''
+        Initializes a new mount point.
+        @source The directory to be mounted.
+        @target The target path in the container to mount at.
+        '''
+
+    @property
+    def source_path(self) -> str:
+        '''
+        @return The directory to be mounted.
+        '''
+
+    @property
+    def mount_path(self) -> str:
+        '''
+        @return The path where the directory should be mounted.
+        '''
+
+
+class IDockerContext(ABC):
+    '''
+    A docker context consists of the necessary data to build a docker image to
+    run a job in.
+    '''
+
+    @abstractmethod
+    def get_dockerfile_source(self) -> str:
+        '''
+        @return The string contents of a Dockerfile which can be used to build
+        the docker image.
+        '''
+
+    @abstractmethod
+    def get_mount_points(self) -> List[MountPoint]:
+        '''
+        @return A list of all mount points for the job.
+        '''
+
+
+class DockerConstraints:
+    '''
+    A list of constraints of the docker container.
+    '''
+    def __init__(self, count_threads: int = -1, memory: int = 1):
+        '''
+        Create a new set of Docker constraints.
+        memory is measured in MB.
```

Comment:
It is explained in the property, do I need to repeat it here?

Reply:
Maybe I can just say that count_thread (after renaming, cpu_threads) is the initial value for the attribute.

Reply:
> do I need to repeat it here

I think all arguments should be documented, even if it's repetitive.
This is because of how the documentation will be used.
Imagine someone who wants to know what arguments the Job constructor takes.
They will typically use their IDE to jump to the constructor definition.
If the constructor does not explain the arguments they would get confused.
Also consider that IDEs typically only process the immediate documentation of a method/constructor to assist the user.